### ISSUE TITLE
trailing-comma rule should not suggest comma after an object rest param in destructuring

### ISSUE THREAD

**Issue by mgol** [palantir/tslint#4172]
### Bug Report

- __TSLint version__: 5.11.0 (also, what the Playground uses)
- __TypeScript version__: 3.1.0-rc (also, what the Playground uses)
- __Running TSLint via__: (pick one) Node.js & Playground

#### TypeScript code being linted

```ts
declare const o;

const {
    a,
    ...rest
} = o;
```

with `tslint.json` configuration:

```json
{
    "rules": {
        "trailing-comma": [true, {"multiline": "always", "singleline": "never"}]
    }
}
```

See it on [Playground](https://palantir.github.io/tslint-playground/?saved=N4Igxg9gJgpiBcJZgDYEMBOMAEkB2AzgC7YQDcAOnlfsdsFdk9mgDSPMB03WxVAvtgC8pSnhCtwEPADMAlgHMEIBnmbYKIGAA8iMPFAKb42ANoBddmuaaMAVxQwjCeh3VNNRDGjko5eBQBaSABbELRjMy87GFZXEBCHIl9-GEjNNBQAdzQAT2c4zQJ-BUc-PDSXTQqANxgMTX5zN2x+ASoQfiA).

#### Actual behavior

An error occurs, indicating a trailing comma is expected.

#### Expected behavior

Latest TypeScript disallows a trailing comma after an object rest param in destructuring to align with ECMAScript. This TSLint behavior requires to apply changes to the code that will make it not compile.

This was raised before in #3147 & #3795. However, then it was only a question of aligning with JavaScript and a style preference. Now it's an actual bug.

**Comment by JoshuaKGoldberg** [comment]
Adding `"esSpecCompliant": true` to the rule's options object fixes this.

Should that be enabled by default?

**Comment by mgol** [comment]
Yes, I think so. That's how I work around it in my code but default settings shouldn't require invalid syntax.

TBH, barring the breaking changes issue, I think this option shouldn't even exist as it's no longer possible to be non-ES compliant anymore here. IIRC this option was added as some people wanted to write code in a syntax compatible with upstream ECMAScript. The TypeScript language tries to remain close to ES syntax, though, so eventually, this hole was closed.

Being able to set `"esSpecCompliant": false` is only relevant to users of older TypeScript versions.

**Comment by adidahiya** [comment]
`esSpecCompliant: true` is enabled in all the built-in configs, `tslint:all`, `tslint:recommended`, and `tslint:latest`, which I believe should be sufficient for most users. we can change the behavior of the rule to make `esSpecCompliant: true` the default behavior in v6.0

**Comment by JoshuaKGoldberg** [comment]
Removing the `Type: Breaking Change` label per #4811. Now accepting PRs!

**Comment by JoshuaKGoldberg** [comment]
### 💀 _**It's time!**_ 💀

**TSLint is deprecated** and no longer accepting pull requests other than security fixes. See #4534. ☠️ 
We recommend you instead use **[typescript-eslint](https://typescript-eslint.io)** to lint your TypeScript code with **ESLint**. ✅

👋 It was a pleasure open sourcing with you!

**Comment by JoshuaKGoldberg** [comment]
🤖 Beep boop! 👉  **TSLint is deprecated** 👈 _(#4534)_  and you should **switch to [typescript-eslint](https://typescript-eslint.io)**! 🤖 

🔒 This issue is being locked to prevent further unnecessary discussions. Thank you! 👋
